tooling/car/unixfs.go

```diff
@@ -164,6 +186,62 @@ func (d *UnixfsDag) MustGetChildrenCids(names ...string) []string {
 	return cids
 }
 
+// MustGetCidsInHAMT returns the cids in the HAMT at the given path. Does not include the CID of the HAMT root
+func (d *UnixfsDag) MustGetCidsInHAMT(names ...string) []string {
+	node := d.MustGetNode(names...)
+	var cids []string
+	tracker := dservTrackingWrapper{
+		DAGService: node.dsvc,
+	}
+
+	lsys := cidlink.DefaultLinkSystem()
+	unixfsnode.AddUnixFSReificationToLinkSystem(&lsys)
+	lsys.StorageReadOpener = func(linkContext linking.LinkContext, link datamodel.Link) (io.Reader, error) {
+		nd, err := tracker.Get(linkContext.Ctx, link.(cidlink.Link).Cid)
+		if err != nil {
+			return nil, err
+		}
+		return bytes.NewReader(nd.RawData()), nil
+	}
+
+	primeNodeBuilder := dagpb.Type.PBNode.NewBuilder()
+	err := dagpb.DecodeBytes(primeNodeBuilder, node.node.RawData())
+	if err != nil {
+		panic(err)
+	}
+	primeNode := primeNodeBuilder.Build()
+	_, err = lsys.KnownReifiers["unixfs-preload"](linking.LinkContext{}, primeNode, &lsys)
+	if err != nil {
+		panic(err)
+	}
+
+	for _, c := range tracker.requestedCids {
+		cids = append(cids, c.String())
+	}
+	return cids
+}
+
+// MustGetCIDsInHAMTTraversal returns the cids needed for a given HAMT traversal. Does not include the HAMT root.
+func (d *UnixfsDag) MustGetCIDsInHAMTTraversal(path []string, child string) []string {
+	node := d.MustGetNode(path...)
+	var cids []string
+	tracker := dservTrackingWrapper{
+		DAGService: node.dsvc,
+	}
+	h, err := hamt.NewHamtFromDag(&tracker, node.node)
+	if err != nil {
+		panic(err)
+	}
```

Comment:
I am concerned how future-proof this is, especially when we have no UnixFS specs yet.
We delegate "what HAMT is" to external library which may change at any time, changing the scope of accepted behaviors.

Let's [copy these explicit asserts](https://github.com/ipfs/boxo/blob/e2fc7f2fd0237afad200d7b0eec8b7a60bdc6644/ipld/unixfs/hamt/hamt.go#LL133C1-L149C3) here, so they are hardcoded here and executed before hamt library:

```suggestion
   pbnd, ok := node.node.(*dag.ProtoNode)
	if !ok {
		return nil, dag.ErrNotProtobuf
	}

	fsn, err := format.FSNodeFromBytes(pbnd.Data())
	if err != nil {
		return nil, err
	}

	if fsn.Type() != format.THAMTShard {
		return nil, fmt.Errorf("node was not a dir shard")
	}

	if fsn.HashType() != HashMurmur3 {
		return nil, fmt.Errorf("only murmur3 supported as hash function")
	}

	h, err := hamt.NewHamtFromDag(&tracker, node.node)
	if err != nil {
		panic(err)
	}
```

Reply:
@lidel the problem with this is that we then have to copy more than the verification: we have to copy many parts from the `Shard` struct defined in the library, in order to perform the remaining operations. Are you sure? That'll likely be quite a bit of code.

Reply:
I'll check

Reply:
[`9f9652f` (#56)](https://github.com/ipfs/gateway-conformance/pull/56/commits/9f9652fa334bf5f404861b4ca0208f3567712368)